### ISSUE TITLE
ShaderGenerator doesn't show TextureStage property changes

### ISSUE THREAD

**Issue by Epihaius** [panda3d/panda3d#177]
While trying to make a luminance slider for a custom color picker, I decided to use multitexturing to decal a semi-transparent black & white gradient texture onto an arbitrarily chosen constant color (changeable by the user). To test out if it would work, I used my own project (Panda3D Studio) to quickly get some results.
Initially it looked like I wanted it, but when I tried to change the color of the first texture stage, the color remained the same. Since I was sure it used to work some time ago, I reverted back to an older version of Panda3D and sure enough, the changes were immediately visible then. It seems to be caused by the ShaderGenerator; when I switch it off, everything works as expected with the newer versions of Panda3D as well.

Here's a small code sample to reproduce the problem:

```
from panda3d.core import *
from direct.showbase.ShowBase import ShowBase


class MyApp(ShowBase):

    def __init__(self):

        ShowBase.__init__(self)

        # comment out the following to see changes made to the first texture stage
        # using the fixed-function pipeline
        self.render.set_shader_auto()

        # Setup the camera

        self.cam.set_pos(-7., -10., 4.)
        self.cam.look_at(0., 0., 0.)

        # Create the texture stages

        self._ts1 = ts1 = TextureStage("flat_color")
        ts1.set_color((1., 0., 0., 1.))
        ts1.set_sort(0)
        # the first texture stage should show a constant color
        ts1.set_combine_rgb(TextureStage.CM_modulate,
                            TextureStage.CS_constant, TextureStage.CO_src_color,
                            TextureStage.CS_previous, TextureStage.CO_src_color)
        self._ts2 = ts2 = TextureStage("luminance")
        ts2.set_sort(1)
        # the second texture stage should allow the constant color to show through
        # a semi-transparent gradient texture
        ts2.set_mode(TextureStage.M_decal)

        # Create the luminance gradient

        w = 20
        h = 256
        h_ = h - 1
        img = PNMImage(w, h, 4)

        for y in range(h):

            c = 1. if y < h_ / 2 else 0.
            a = 1. - 2. * y / h_ if y < h_ / 2 else 2. * y / h_ - 1.

            for x in range(w):
                img.set_xel(x, y, c, c, c)
                img.set_alpha(x, y, a)

        # Create the geometry and add textures

        cm = CardMaker("lum_quad")
        cm.set_frame(-.5, .5, -2., 2.)
        quad = self.render.attach_new_node(cm.generate())
        tex = Texture("luminance")
        tex.load(img)
        quad.set_texture(self._ts1, Texture(""))
        quad.set_texture(self._ts2, tex)

        self.accept("r", lambda: self.change_color((1., 0., 0., 1.)))
        self.accept("g", lambda: self.change_color((0., 1., 0., 1.)))
        self.accept("b", lambda: self.change_color((0., 0., 1., 1.)))

    def change_color(self, color):

        self._ts1.set_color(color)


app = MyApp()
app.run()
```

When you run the code, you should see a rectangle with a gradient going from white at the top to red in the middle, to black at the bottom. When you press 'R', 'G' or 'B' on the keyboard, the middle color on the rectangle should change to red, green or blue respectively. Without the ShaderGenerator it should always work, but with newer Panda3D versions it doesn't work when the ShaderGenerator is used.

Other property changes of the first texture stage did not show up either, e.g. combine mode parameters.

There were no problems with Panda3D version 560fd4e (2017-05-13), but from version f79fbf2 (2017-07-10) onward the issue appeared (I did not try other versions in between).

This happens on Windows 8.1 64-bit.

**Comment by rdb** [comment]
This was an oversight, and should be easy to fix.  I'll check in a fix soon.  In the meantime, as a workaround, you can use the `CS_constant_color_scale` mode and change the color scale instead.

Allowing the other combine mode, sources, and operands to be changed at runtime is slightly trickier.  Do you also need this?  This wasn't supported mostly because it seemed unlikely to be useful.

**Comment by Epihaius** [comment]
> This was an oversight, and should be easy to fix. I'll check in a fix soon.

Ah, that will be great, thank you!

> In the meantime, as a workaround, you can use the CS_constant_color_scale mode and change the color scale instead.

Hmmm, that works for the lower, black portion of the gradient texture, but the white upper half gets completely colorized as well, so I guess I'll wait for the fix.

> Allowing the other combine mode, sources, and operands to be changed at runtime is slightly trickier. Do you also need this? This wasn't supported mostly because it seemed unlikely to be useful.

Indeed, for most purposes it probably wouldn't be very useful, but, you know: Panda3D Studio ;) .
The whole idea of that project is to provide a sandbox-like editor, allowing its users to experiment with different values for different settings on-the-fly, so they can easily finetune the models/materials/textures they're working on.
Thanks to being able to change the properties of a texture stage directly, this was rather straightforward to implement. If that is no longer possible, then I'll just have to reapply the textures using copies of the changed texture stages on all of the affected models. Something similar already has to happen for changes in material properties anyway, so I should be able to do it, even if it's less convenient.

**Comment by rdb** [comment]
OK, gotcha.  Can you make a test case and file a new issue for the other combine properties?  Be sure to include whether it also occurs for non-combine properties (like `set_mode`).

**Comment by Epihaius** [comment]
Thank you in advance!

A new issue has been filed for the remaining properties - and it looks like all of them are affected; e.g. `set_mode` has no effect, neither has `set_rgb_scale`.

**Comment by Epihaius** [comment]
A while ago I filed another issue related to texture stages [on Launchpad](https://bugs.launchpad.net/panda3d/+bug/1693835) and in case it would get overlooked there, I copied it here as issue #179

**Comment by rdb** [comment]
> Something similar already has to happen for changes in material properties anyway, so I should be able to do it, even if it's less convenient.

Which material properties did this concern?  I thought this shouldn't be an issue for material properties as of f79fbf2, at least.  If not, I might be able to fix this as part of #178.

**Comment by Epihaius** [comment]
Oh I didn't realize that material attributes no longer get locked; this is great!
I just tested and everything works just fine, thank you!
